hdmi/models/decoder_model.py

```diff
@@ -0,0 +1,265 @@
+         This is modelled after the xapp495 decoder module.
+
+        Args:
+            clock: The system clock or the pixel clock
+            data_in: The TMDS data (10 bits width) to be decoded
+            video_preamble: signal to detect the video preamble in the input data
+            data_island_preamble: signal to detect the data island preamble in the input data
+            c0: Control signal (hsync for Blue channel)
+            c1: Control signal (vsync for Blue channel)
+            vde: Video Data enable
+            ade: Audio data enable
+            video_out: Output video data
+            audio_out: Output audio (or aux) data.
+            channel: Color of the channel ('RED', 'GREEN' or 'BLUE'). Default value is 'BLUE'
+
+        Usage:
+            decoder_model = DecoderModel(*params)
+            process_inst = decoder_model.process()
+            process_inst.run_sim()
+
+        """
+
+        self.clock = clock
+        self.data_in = data_in
+        self.video_preamble = video_preamble
+        self.data_island_preamble = data_island_preamble
+        self.c0 = c0
+        self.c1 = c1
+        self.vde = vde
+        self.ade = ade
+        self.video_out = video_out
+        self.audio_out = audio_out
+        self.channel = channel
+
+    def get_video_data(self):
+
+        """
+
+        Returns:
+            The output video data
+
+        """
+
+        return self.video_out
+
+    def get_audio_data(self):
+
+        """
+
+        Returns:
+            The output audio data(or AUX data)
+
+        """
+
+        return self.audio_out
+
+    def read(self):
+
+        """
+
+        Waits for a positive edge of clock
+
+        """
+
+        yield self.clock.posedge
+
+    def write_data(self, data_in):
+
+        """
+
+        Writes the given data onto the input data signal
+
+        Args:
+            data_in: 10 bit intbv value(or a 10 bit integer)
+
+        """
+
+        self.data_in.next = data_in
+        yield self.clock.posedge
+
+
+    @block
+    def process(self):
+
+        """
+
+        It simulates the decoding process of the TMDS decoder.
+
+        Usage:
+            process_inst = decoder_model.process()
+            process_inst.run_sim()
+
+        """
+
+        control_token = [int('1101010100', 2),  # 00
+                         int('0010101011', 2),  # 01
+                         int('0101010100', 2),  # 10
+                         int('1010101011', 2)]  # 11
+
+        # Control signals
+        control, _control, control_end = [Signal(bool(0)) for _ in range(3)]
+
+        # Signals to detect video period and data island period
+        video_period = Signal(bool(0))
+        data_island_period = Signal(bool(0))
+
+        data = Signal(intbv(0)[8:0])
+
+        @always_comb
+        def continuous_assignment():
+            data.next = ~self.data_in[8:0] if self.data_in[9] == 1 else self.data_in[8:0]
```

Comment:
This line is a little long, probably should use the multiline `if ... else` instead of the conditional oneline.